EIPS/eip-7928.md

```diff
@@ -0,0 +1,199 @@
+
+## Specification
+
+### Block Structure Modification
+
+We introduce three new components in the block body:
+
+1. A Block Access List (BAL) that maps transactions to accessed addresses and storage slots, transaction indices, and post-execution values for writes.
+2. Balance Diffs that track every address touched by value transfers along with the balance deltas.
+3. Nonces that record the pre-block nonces of all transaction senders within the block.
+
+### SSZ Data Structures
+
+```python
+# Type aliases
+Address = ByteVector(20)
+StorageKey = ByteVector(32)
+StorageValue = ByteVector(32)
+TxIndex = uint16
+Nonce = uint64
+
+# Constants; chosen to support a 630m block gas limit
+MAX_TXS = 30_000
+MAX_SLOTS = 300_000
+MAX_ACCOUNTS = 300_000
+
+# SSZ containers
+class PerTxAccess(Container):
+    tx_index: TxIndex
+    value_after: StorageValue # value in state after the last access
+
+class SlotAccess(Container):
+    slot: StorageKey
+    accesses: List[PerTxAccess, MAX_TXS] # empty for reads
+
+class AccountAccess(Container):
+    address: Address
+    accesses: List[SlotAccess, MAX_SLOTS]
+
+BlockAccessList = List[AccountAccess, MAX_ACCOUNTS]
+
+# Balance Diff structures
+BalanceDelta = ByteVector(11)  # signed, two's complement encoding
+
+class BalanceChange(Container):
+    tx_index: TxIndex
+    delta: BalanceDelta  # signed integer, encoded as 11-byte vector
+
+class AccountBalanceDiff(Container):
+    address: Address
+    changes: List[BalanceChange, MAX_TXS]
+
+BalanceDiffs = List[AccountBalanceDiff, MAX_ACCOUNTS]
+
+# Post-block nonce structures
+class AccountNonce(Container):
+    address: Address  # account address
+    nonce_after: Nonce  # nonce value after block execution
+
+NonceDiffs = List[AccountNonceDiff, MAX_TXS]
+```
+
+The `BlockAccessList` is a deduplicated list of accessed addresses. For each address, it MUST contain a list of accessed storage keys. 
+For writes, each `SlotAccess` MUST contain an ordered list of transaction indices that accessed this key, and the final storage value after the last access. Transactions with writes that do not change the storage value MUST NOT contain a `value_after`.
+For contract deployments, the `value_after` MUST contain the runtime bytecode of the newly deployed contract.
+
+For reads, each `SlotAccess` MUST contain an empty `accesses` list.
+
+The `BalanceDiffs` structure tracks every address with a balance change, including transaction senders, recipients, and the block's coinbase address. 
+Each entry MUST include the transaction index and the signed balance delta per address for each transaction.
+11 bytes are sufficient to represent the total ETH supply.
+
+The `NonceDiffs` structure MUST record the post-block nonce values for all `CREATE2` deployer accounts and the deploayed contracts in the block.
+
+### State Transition Function
+
+Modify the state transition function to validate the block-level access lists:
+
+```python 
+def state_transition(block: Block) -> None:
```

Comment:
Looks fine but we can likely take out a lot of items for default behavior. (Of course we should still account for the worst case scenario).